lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp

```diff
@@ -0,0 +1,770 @@
+  static ConstString kld_off_next_symbol_name("kld_off_next");
+  static ConstString kld_off_filename_symbol_name("kld_off_filename");
+  static ConstString kld_off_pathname_symbol_name("kld_off_pathname");
+  const Symbol *kld_off_address_symbol =
+      m_kernel_image_info.GetModule()->FindFirstSymbolWithNameAndType(
+          kld_off_address_symbol_name, eSymbolTypeData);
+  const Symbol *kld_off_next_symbol =
+      m_kernel_image_info.GetModule()->FindFirstSymbolWithNameAndType(
+          kld_off_next_symbol_name, eSymbolTypeData);
+  const Symbol *kld_off_filename_symbol =
+      m_kernel_image_info.GetModule()->FindFirstSymbolWithNameAndType(
+          kld_off_filename_symbol_name, eSymbolTypeData);
+  const Symbol *kld_off_pathname_symbol =
+      m_kernel_image_info.GetModule()->FindFirstSymbolWithNameAndType(
+          kld_off_pathname_symbol_name, eSymbolTypeData);
+
+  if (!kld_off_address_symbol || !kld_off_next_symbol ||
+      !kld_off_filename_symbol || !kld_off_pathname_symbol)
+    return false;
+  Status error;
+  const int32_t kld_off_address = m_process->ReadSignedIntegerFromMemory(
+      kld_off_address_symbol->GetAddress().GetLoadAddress(
+          &m_process->GetTarget()),
+      4, 0, error);
+  if (error.Fail())
+    return false;
+  const int32_t kld_off_next = m_process->ReadSignedIntegerFromMemory(
+      kld_off_next_symbol->GetAddress().GetLoadAddress(&m_process->GetTarget()),
+      4, 0, error);
+  if (error.Fail())
+    return false;
+  const int32_t kld_off_filename = m_process->ReadSignedIntegerFromMemory(
+      kld_off_filename_symbol->GetAddress().GetLoadAddress(
+          &m_process->GetTarget()),
+      4, 0, error);
+  if (error.Fail())
+    return false;
+
+  const int32_t kld_off_pathname = m_process->ReadSignedIntegerFromMemory(
+      kld_off_pathname_symbol->GetAddress().GetLoadAddress(
+          &m_process->GetTarget()),
+      4, 0, error);
+  if (error.Fail())
+    return false;
+
+  // Parse KMods
+  addr_t kld_load_addr(LLDB_INVALID_ADDRESS);
+  char kld_filename[255];
+  char kld_pathname[255];
+  addr_t current_kld =
+      linker_files_head_addr.GetLoadAddress(&m_process->GetTarget());
+
+  while (current_kld != 0) {
+    m_process->ReadCStringFromMemory(
+        m_process->ReadPointerFromMemory(current_kld + kld_off_filename, error),
+        kld_filename, sizeof(kld_filename), error);
+    m_process->ReadCStringFromMemory(
+        m_process->ReadPointerFromMemory(current_kld + kld_off_pathname, error),
+        kld_pathname, sizeof(kld_pathname), error);
+    kld_load_addr =
+        m_process->ReadPointerFromMemory(current_kld + kld_off_address, error);
+
+    kmods_list.emplace_back();
+    KModImageInfo &kmod_info = kmods_list.back();
+    kmod_info.SetName(kld_filename);
+    kmod_info.SetLoadAddress(kld_load_addr);
+    kmod_info.SetPath(kld_pathname);
+    current_kld =
+        m_process->ReadPointerFromMemory(current_kld + kld_off_next, error);
```

Comment:
Shouldn't you check the result of those memory reads above before mutating state here? Otherwise you may end up with stale or otherwise incorrect information, no?

Reply:
> Shouldn't you check the result of those memory reads above before mutating state here? Otherwise you may end up with stale or otherwise incorrect information, no?

This will return LLDB_INVALID_ADDRESS if it fails, so you can check the error if you want, or just need to deal with LLDB_INVALID_ADDRESS correctly.

Reply:
But yes, check if you got valid strings back for "kld_filename", "kld_pathname" and "kld_load_addr != LLDB_INVALID_ADDRESS would be good idea before adding the to the kmods_list. The "current_kld" should be checked against zero and LLDB_INVALID_ADDRESS in the while loop

Reply:
Can I just check the error that was passed into the function?

Reply:
Either one is fine, you could also do both? This is one place where LLDB can be a little frustrating because there are 2 ways to detect the error and it's not obvious which one you should be checking.